cpp/src/arrow/array/builder_dict.h

```diff
@@ -93,6 +93,10 @@ class ARROW_EXPORT DictionaryMemoTable {
   std::unique_ptr<DictionaryMemoTableImpl> impl_;
 };
 
+/// \brief Check array's value type by DCHECK
+ARROW_EXPORT void CheckArrayType(const std::shared_ptr<DataType>& expected_type,
+                                 const Array& array, const std::string& message);
```

Comment:
Can you make this a protected method of `DictionaryBuilderBase` or something? We don't want to expose it as a public API.

Reply:
Also, take `const char*` for the error message, to avoid constructing a `std::string`.

Reply:
I'll make this a protected method of `ArrayBuilder`.